### PR TITLE
Setup branch protection for set/main

### PR DESCRIPTION
**Pull request by TruongQuangSB**
This PR setup branch protection for main branch of all eclipse-set repo:
- Allow only commit over PR (except eclipse-set-bot)
- Disable force push
- Don't allow delete branch

### CODE REVIEW

File: otterdog/eclipse-set.jsonnet
Comment (netomi) [inline otterdog/eclipse-set.jsonnet]: you should use predefined functions to create resources as this would take default values into account. See an example here: https://otterdog.readthedocs.io/en/latest/reference/organization/repository/branch-protection-rule/

Comment (netomi) [reply otterdog/eclipse-set.jsonnet]: Another real world example can be seen here: https://github.com/adoptium/.eclipsefdn/blob/main/otterdog/adoptium.jsonnet#L271

**Comment by netomi** [comment]
ty for the PR. Please see my comments about using the respective function orgs.newBranchProtectionRule('<pattern>') and then overriding the settings that you would like to have.

A list of default values for the function can be found here: https://eclipse-set.github.io/.eclipsefdn/

btw. I noticed that the project lead of the project (Christoph Klaus) does not have a github handle associated to his eclipse account and is also not the member of the organization. Is this expected? Who is going to approve changes to the infrastructure?

File: otterdog/eclipse-set.jsonnet
Comment (TruongQuangSB) [reply otterdog/eclipse-set.jsonnet]: ok, thank you for the tip

File: otterdog/eclipse-set.jsonnet
Comment (netomi) [inline otterdog/eclipse-set.jsonnet]: looks much better now.

The 3 settings:

- requires_pull_request
- allows_deletions
- allows_force_pushes

could be removed, as the same values are already inherited from the default config.

**Comment by netomi** [comment]
Also please take a look at the final configuration (look at the comment from the github actions bot and expand the message). It will show the full settings that will be used for the branch protection rule, taking into account all values from the default config.

e.g. number of required approvals is set to 2. This might be too high for some projects especially if it does not have a high number of active committers atm.

File: otterdog/eclipse-set.jsonnet
Comment (TruongQuangSB) [reply otterdog/eclipse-set.jsonnet]: ok, hope everythings ok this time :)

File: otterdog/eclipse-set.jsonnet
Comment (netomi) [inline otterdog/eclipse-set.jsonnet]: if you want to disable approvals but still require pull requests, you need to set this to property to 0

you should only set this to null if requires_pull_request is false. Its a bit confusing, but we follow here the API from GitHub as much as possible.

Comment (TruongQuangSB) [reply otterdog/eclipse-set.jsonnet]: ah ok

**Comment by netomi** [comment]
For me the PR is good to be merged, just waiting for an approval from at least another committer.

**Review by netomi — APPROVED** [review]

**Review by pstuecker — CHANGES_REQUESTED** [review]
I think we should apply this to all repositories, rather than just the main SET repository. Change itself looks fine otherwise.

**Comment by pstuecker** [comment]
> btw. I noticed that the project lead of the project (Christoph Klaus) does not have a github handle associated to his eclipse account and is also not the member of the organization. Is this expected? 

@netomi 
I guess he hasn't added his Github account yet. I'll send him an email. As he isn't immediately involved in the Java development of SET, him not being around wasn't an issue yet (as we are also still working on setting up some Github things). 

> Who is going to approve changes to the infrastructure?

That's a good question. What are the requirements from the EF regarding this? For us the ideal situation would likely be to have us committers be able to approve it (with multiple approving reviews).

**Comment by netomi** [comment]
@pstuecker Personally I think it would be good if the project lead is at least aware of the things that are going on. Regarding the requirements from the EF: in general it is up to the project to decide how they want such changes to be approved. If there are a small number of committers and you trust all of them, having an approval from another committer is certainly sufficient imho, especially if the project is still in incubation phase and has to setup a lot of things. An initial approval on this process from the project lead would be good imho.

**Comment by TruongQuangSB** [comment]
> I think we should apply this to all repositories, rather than just the main SET repository. Change itself looks fine otherwise.

I have apply for all repo

**Comment by netomi** [comment]
> > I think we should apply this to all repositories, rather than just the main SET repository. Change itself looks fine otherwise.
> 
> I have apply for all repo

very nice way to apply the same rule to all repos.

May I suggest to slightly change the function to that:

```
local custom_branch_protection_rule(pattern) =
  orgs.newBranchProtectionRule(pattern) {
    restricts_pushes: true,
    push_restrictions: [ '@eclipse-set-bot' ],
    required_approving_review_count: 0,
  };
```

so that it also takes the pattern into account.

**Comment by TruongQuangSB** [comment]
> > > I think we should apply this to all repositories, rather than just the main SET repository. Change itself looks fine otherwise.
> > 
> > 
> > I have apply for all repo
> 
> very nice way to apply the same rule to all repos.
> 
> May I suggest to slightly change the function to that:
> 
> ```
> local custom_branch_protection_rule(pattern) =
>   orgs.newBranchProtectionRule(pattern) {
>     restricts_pushes: true,
>     push_restrictions: [ '@eclipse-set-bot' ],
>     required_approving_review_count: 0,
>   };
> ```
> 
> so that it also takes the pattern into account.

ok, this ist better :)

**Comment by netomi** [comment]
> > > > I think we should apply this to all repositories, rather than just the main SET repository. Change itself looks fine otherwise.
> > > 
> > > 
> > > I have apply for all repo
> > 
> > 
> > very nice way to apply the same rule to all repos.
> > May I suggest to slightly change the function to that:
> > ```
> > local custom_branch_protection_rule(pattern) =
> >   orgs.newBranchProtectionRule(pattern) {
> >     restricts_pushes: true,
> >     push_restrictions: [ '@eclipse-set-bot' ],
> >     required_approving_review_count: 0,
> >   };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > so that it also takes the pattern into account.
> 
> ok, this ist better :)

you will have to quote the pattern, like ('main')

**Review by pstuecker — APPROVED** [review]

**Comment by TruongQuangSB** [comment]
> > > > > I think we should apply this to all repositories, rather than just the main SET repository. Change itself looks fine otherwise.
> > > > 
> > > > 
> > > > I have apply for all repo
> > > 
> > > 
> > > very nice way to apply the same rule to all repos.
> > > May I suggest to slightly change the function to that:
> > > ```
> > > local custom_branch_protection_rule(pattern) =
> > >   orgs.newBranchProtectionRule(pattern) {
> > >     restricts_pushes: true,
> > >     push_restrictions: [ '@eclipse-set-bot' ],
> > >     required_approving_review_count: 0,
> > >   };
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > so that it also takes the pattern into account.
> > 
> > 
> > ok, this ist better :)
> 
> you will have to quote the pattern, like ('main')

**Comment by netomi** [comment]
LGTM, I will approve and merge this PR for now, but please try to get Christoph Klaus on board of the organization.

**Comment by netomi** [comment]
change has been applied, branch protection rules have been created.

**Pull request merged**
